[ICE]: delegation: `No HirId for DefId`

<!--
[31mICE[0m: Rustc ./a.rs '' 'error: internal compiler error: compiler/rustc_middle/src/hir/mod.rs:478:32: No HirId for DefId(0:4 ~ a[c659]::foo::{constant#0})', 'error: internal compiler error: compiler/rustc_middle/src/hir/mod.rs:478:32: No HirId for DefId(0:4 ~ a[c659]::foo::{constant#0})'
File: /tmp/im6/a.rs
-->

auto-reduced (treereduce-rust):
````rust
reuse foo:: < {
    fn foo() {}
    reuse foo;
  }
  >;

````

original:
````rust
#![feature(fn_delegation)]

reuse foo:: < {
    fn foo() {}
    reuse foo;
  }
  >;

````

Version information
````
rustc 1.96.0-nightly (981cf6947 2026-04-04)
binary: rustc
commit-hash: 981cf69479ded5e2de0cf9e16111c19d65be0369
commit-date: 2026-04-04
host: x86_64-unknown-linux-gnu
release: 1.96.0-nightly
LLVM version: 22.1.2

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/981cf69479ded5e2de0cf9e16111c19d65be0369/compiler/rustc_middle/src/hir/mod.rs#L472-L484

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0428]: the name `foo` is defined multiple times
 --> /tmp/icemaker_global_tempdir.B6BTH9LlK7Uj/rustc_testrunner_tmpdir_reporting.VMLpdNPqpxHZ/mvce.rs:3:5
  |
2 |     fn foo() {}
  |     -------- previous definition of the value `foo` here
3 |     reuse foo;
  |     ^^^^^^^^^^ `foo` redefined here
  |
  = note: `foo` must be defined only once in the value namespace of this block

error[E0658]: functions delegation is not yet fully implemented
 --> /tmp/icemaker_global_tempdir.B6BTH9LlK7Uj/rustc_testrunner_tmpdir_reporting.VMLpdNPqpxHZ/mvce.rs:3:5
  |
3 |     reuse foo;
  |     ^^^^^^^^^^
  |
  = note: see issue #118212 <https://github.com/rust-lang/rust/issues/118212> for more information
  = help: add `#![feature(fn_delegation)]` to the crate attributes to enable
  = note: this compiler was built on 2026-04-04; consider upgrading it if it is out of date

error[E0658]: functions delegation is not yet fully implemented
 --> /tmp/icemaker_global_tempdir.B6BTH9LlK7Uj/rustc_testrunner_tmpdir_reporting.VMLpdNPqpxHZ/mvce.rs:1:1
  |
1 | / reuse foo:: < {
2 | |     fn foo() {}
3 | |     reuse foo;
4 | |   }
5 | |   >;
  | |____^
  |
  = note: see issue #118212 <https://github.com/rust-lang/rust/issues/118212> for more information
  = help: add `#![feature(fn_delegation)]` to the crate attributes to enable
  = note: this compiler was built on 2026-04-04; consider upgrading it if it is out of date

error: failed to resolve delegation callee
 --> /tmp/icemaker_global_tempdir.B6BTH9LlK7Uj/rustc_testrunner_tmpdir_reporting.VMLpdNPqpxHZ/mvce.rs:1:7
  |
1 | reuse foo:: < {
  |       ^^^

error: internal compiler error: /rustc-dev/981cf69479ded5e2de0cf9e16111c19d65be0369/compiler/rustc_middle/src/hir/mod.rs:478:32: No HirId for DefId(0:4 ~ mvce[ce0a]::foo::{constant#0})


thread 'rustc' (3346370) panicked at /rustc-dev/981cf69479ded5e2de0cf9e16111c19d65be0369/compiler/rustc_middle/src/hir/mod.rs:478:32:
Box<dyn Any>
stack backtrace:
   0:     0x7fd7519affbb - <<std[ebbe8e490f6d8018]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[e0039014fbc551a9]::fmt::Display>::fmt
   1:     0x7fd75201e1c8 - core[e0039014fbc551a9]::fmt::write
   2:     0x7fd7519c6fc6 - <std[ebbe8e490f6d8018]::sys::stdio::unix::Stderr as std[ebbe8e490f6d8018]::io::Write>::write_fmt
   3:     0x7fd751985fc8 - std[ebbe8e490f6d8018]::panicking::default_hook::{closure#0}
   4:     0x7fd7519a33c3 - std[ebbe8e490f6d8018]::panicking::default_hook
   5:     0x7fd7509ada9c - std[ebbe8e490f6d8018]::panicking::update_hook::<alloc[cb39f600d6f52ef4]::boxed::Box<rustc_driver_impl[c03a371dcba30999]::install_ice_hook::{closure#1}>>::{closure#0}
   6:     0x7fd7519a36a2 - std[ebbe8e490f6d8018]::panicking::panic_with_hook
   7:     0x7fd7509dd3a1 - std[ebbe8e490f6d8018]::panicking::begin_panic::<rustc_errors[24e2b13995366241]::ExplicitBug>::{closure#0}
   8:     0x7fd7509d6086 - std[ebbe8e490f6d8018]::sys::backtrace::__rust_end_short_backtrace::<std[ebbe8e490f6d8018]::panicking::begin_panic<rustc_errors[24e2b13995366241]::ExplicitBug>::{closure#0}, !>
   9:     0x7fd7509d5deb - std[ebbe8e490f6d8018]::panicking::begin_panic::<rustc_errors[24e2b13995366241]::ExplicitBug>
  10:     0x7fd7509e89e1 - <rustc_errors[24e2b13995366241]::diagnostic::BugAbort as rustc_errors[24e2b13995366241]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  11:     0x7fd750fbfb49 - rustc_middle[a5faf31693dbe3fb]::util::bug::opt_span_bug_fmt::<rustc_span[e72d8c7181f2d66e]::span_encoding::Span>::{closure#0}
  12:     0x7fd750fbfcd2 - rustc_middle[a5faf31693dbe3fb]::ty::context::tls::with_opt::<rustc_middle[a5faf31693dbe3fb]::util::bug::opt_span_bug_fmt<rustc_span[e72d8c7181f2d66e]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  13:     0x7fd750fb028b - rustc_middle[a5faf31693dbe3fb]::ty::context::tls::with_context_opt::<rustc_middle[a5faf31693dbe3fb]::ty::context::tls::with_opt<rustc_middle[a5faf31693dbe3fb]::util::bug::opt_span_bug_fmt<rustc_span[e72d8c7181f2d66e]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  14:     0x7fd74e10a2f4 - rustc_middle[a5faf31693dbe3fb]::util::bug::bug_fmt
  15:     0x7fd7520bc313 - rustc_query_impl[77c21dd70d39b376]::query_impl::local_def_id_to_hir_id::invoke_provider_fn::__rust_begin_short_backtrace
  16:     0x7fd7520bb626 - rustc_query_impl[77c21dd70d39b376]::execution::try_execute_query::<rustc_data_structures[557cea3a9825f03a]::vec_cache::VecCache<rustc_span[e72d8c7181f2d66e]::def_id::LocalDefId, rustc_middle[a5faf31693dbe3fb]::query::erase::ErasedData<[u8; 8usize]>, rustc_middle[a5faf31693dbe3fb]::dep_graph::graph::DepNodeIndex>, false>
  17:     0x7fd7520bb385 - rustc_query_impl[77c21dd70d39b376]::query_impl::local_def_id_to_hir_id::execute_query_non_incr::__rust_end_short_backtrace
  18:     0x7fd7521cc74c - <rustc_span[e72d8c7181f2d66e]::def_id::LocalDefId as rustc_hir[cbdb9c4d50c2ed1d]::attrs::HasAttrs<rustc_middle[a5faf31693dbe3fb]::ty::context::TyCtxt>>::get_attrs
  19:     0x7fd7521cbcbb - rustc_query_impl[77c21dd70d39b376]::query_impl::sanitizer_settings_for::invoke_provider_fn::__rust_begin_short_backtrace
  20:     0x7fd7521cb0fb - rustc_query_impl[77c21dd70d39b376]::execution::try_execute_query::<rustc_data_structures[557cea3a9825f03a]::vec_cache::VecCache<rustc_span[e72d8c7181f2d66e]::def_id::LocalDefId, rustc_middle[a5faf31693dbe3fb]::query::erase::ErasedData<[u8; 4usize]>, rustc_middle[a5faf31693dbe3fb]::dep_graph::graph::DepNodeIndex>, false>
  21:     0x7fd7521cae7b - rustc_query_impl[77c21dd70d39b376]::query_impl::sanitizer_settings_for::execute_query_non_incr::__rust_end_short_backtrace
  22:     0x7fd7521cbd2f - rustc_query_impl[77c21dd70d39b376]::query_impl::sanitizer_settings_for::invoke_provider_fn::__rust_begin_short_backtrace
  23:     0x7fd7521cb0fb - rustc_query_impl[77c21dd70d39b376]::execution::try_execute_query::<rustc_data_structures[557cea3a9825f03a]::vec_cache::VecCache<rustc_span[e72d8c7181f2d66e]::def_id::LocalDefId, rustc_middle[a5faf31693dbe3fb]::query::erase::ErasedData<[u8; 4usize]>, rustc_middle[a5faf31693dbe3fb]::dep_graph::graph::DepNodeIndex>, false>
  24:     0x7fd7521cae7b - rustc_query_impl[77c21dd70d39b376]::query_impl::sanitizer_settings_for::execute_query_non_incr::__rust_end_short_backtrace
  25:     0x7fd752580a28 - rustc_codegen_ssa[734e7c4a27baa15c]::codegen_attrs::codegen_fn_attrs
  26:     0x7fd75257f162 - rustc_query_impl[77c21dd70d39b376]::query_impl::codegen_fn_attrs::invoke_provider_fn::__rust_begin_short_backtrace
  27:     0x7fd75205ef0a - rustc_query_impl[77c21dd70d39b376]::execution::try_execute_query::<rustc_middle[a5faf31693dbe3fb]::query::caches::DefIdCache<rustc_middle[a5faf31693dbe3fb]::query::erase::ErasedData<[u8; 8usize]>>, false>
  28:     0x7fd75205e790 - rustc_query_impl[77c21dd70d39b376]::query_impl::codegen_fn_attrs::execute_query_non_incr::__rust_end_short_backtrace
  29:     0x7fd75044f915 - <rustc_ast_lowering[51b7b24872d13be5]::LoweringContext<rustc_ast_lowering[51b7b24872d13be5]::ResolverDelayedAstLowering>>::lower_delegation
  30:     0x7fd7504691b2 - <rustc_ast_lowering[51b7b24872d13be5]::LoweringContext<rustc_ast_lowering[51b7b24872d13be5]::ResolverDelayedAstLowering>>::lower_item_kind
  31:     0x7fd750496647 - <rustc_ast_lowering[51b7b24872d13be5]::item::ItemLowerer<rustc_ast_lowering[51b7b24872d13be5]::ResolverDelayedAstLowering>>::lower_node
  32:     0x7fd75044b429 - rustc_ast_lowering[51b7b24872d13be5]::lower_delayed_owner
  33:     0x7fd7520ba522 - rustc_query_impl[77c21dd70d39b376]::execution::try_execute_query::<rustc_data_structures[557cea3a9825f03a]::vec_cache::VecCache<rustc_span[e72d8c7181f2d66e]::def_id::LocalDefId, rustc_middle[a5faf31693dbe3fb]::query::erase::ErasedData<[u8; 0usize]>, rustc_middle[a5faf31693dbe3fb]::dep_graph::graph::DepNodeIndex>, false>
  34:     0x7fd75144f1db - rustc_query_impl[77c21dd70d39b376]::query_impl::lower_delayed_owner::execute_query_non_incr::__rust_end_short_backtrace
  35:     0x7fd75338d1a5 - rustc_middle[a5faf31693dbe3fb]::hir::map::hir_crate_items
  36:     0x7fd753239425 - rustc_query_impl[77c21dd70d39b376]::query_impl::hir_crate_items::invoke_provider_fn::__rust_begin_short_backtrace
  37:     0x7fd753235725 - rustc_query_impl[77c21dd70d39b376]::execution::try_execute_query::<rustc_middle[a5faf31693dbe3fb]::query::caches::SingleCache<rustc_middle[a5faf31693dbe3fb]::query::erase::ErasedData<[u8; 8usize]>>, false>
  38:     0x7fd75323458f - rustc_query_impl[77c21dd70d39b376]::query_impl::hir_crate_items::execute_query_non_incr::__rust_end_short_backtrace
  39:     0x7fd7520b7c18 - rustc_interface[29be6a7f2a5c93ab]::passes::analysis
  40:     0x7fd753236409 - rustc_query_impl[77c21dd70d39b376]::execution::try_execute_query::<rustc_middle[a5faf31693dbe3fb]::query::caches::SingleCache<rustc_middle[a5faf31693dbe3fb]::query::erase::ErasedData<[u8; 0usize]>>, false>
  41:     0x7fd753236075 - rustc_query_impl[77c21dd70d39b376]::query_impl::analysis::execute_query_non_incr::__rust_end_short_backtrace
  42:     0x7fd7531e5a9e - rustc_interface[29be6a7f2a5c93ab]::interface::run_compiler::<(), rustc_driver_impl[c03a371dcba30999]::run_compiler::{closure#0}>::{closure#1}
  43:     0x7fd7532154fe - std[ebbe8e490f6d8018]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[29be6a7f2a5c93ab]::util::run_in_thread_with_globals<rustc_interface[29be6a7f2a5c93ab]::util::run_in_thread_pool_with_globals<rustc_interface[29be6a7f2a5c93ab]::interface::run_compiler<(), rustc_driver_impl[c03a371dcba30999]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  44:     0x7fd753215da0 - <std[ebbe8e490f6d8018]::thread::lifecycle::spawn_unchecked<rustc_interface[29be6a7f2a5c93ab]::util::run_in_thread_with_globals<rustc_interface[29be6a7f2a5c93ab]::util::run_in_thread_pool_with_globals<rustc_interface[29be6a7f2a5c93ab]::interface::run_compiler<(), rustc_driver_impl[c03a371dcba30999]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[e0039014fbc551a9]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  45:     0x7fd753216cac - <std[ebbe8e490f6d8018]::sys::thread::unix::Thread>::new::thread_start
  46:     0x7fd74cca597a - <unknown>
  47:     0x7fd74cd292bc - <unknown>
  48:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.96.0-nightly (981cf6947 2026-04-04) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [local_def_id_to_hir_id] getting HIR ID of `foo::{constant#0}`
#1 [sanitizer_settings_for] checking what set of sanitizers are enabled on `foo::{constant#0}`
#2 [sanitizer_settings_for] checking what set of sanitizers are enabled on `foo::{constant#0}::foo`
#3 [codegen_fn_attrs] computing codegen attributes of `foo::{constant#0}::foo`
#4 [lower_delayed_owner] lowering the delayed AST owner `foo::{constant#0}::foo`
#5 [hir_crate_items] getting HIR crate items
#6 [analysis] running analysis passes on crate `mvce`
end of query stack
error: aborting due to 5 previous errors

Some errors have detailed explanations: E0428, E0658.
For more information about an error, try `rustc --explain E0428`.

```

</p>
</details>

<!--
query stack:
#0 [local_def_id_to_hir_id] getting HIR ID of `foo::{constant#0}`
#1 [sanitizer_settings_for] checking what set of sanitizers are enabled on `foo::{constant#0}`
#2 [sanitizer_settings_for] checking what set of sanitizers are enabled on `foo::{constant#0}::foo`
#3 [codegen_fn_attrs] computing codegen attributes of `foo::{constant#0}::foo`
#4 [lower_delayed_owner] lowering the delayed AST owner `foo::{constant#0}::foo`
#5 [hir_crate_items] getting HIR crate items
#6 [analysis] running analysis passes on crate `mvce`

-->
@rustbot label +F-fn_delegation

--- a/compiler/rustc_ast_lowering/src/delegation.rs
+++ b/compiler/rustc_ast_lowering/src/delegation.rs
@@ -128,14 +128,12 @@ impl<'hir, R: ResolverAstLoweringExt<'hir>> LoweringContext<'_, 'hir, R> {
         {
             self.get_sig_id(delegation_info.resolution_node, span)
         } else {
-            return self.generate_delegation_error(
-                self.dcx().span_delayed_bug(
-                    span,
-                    format!("LoweringContext: the delegation {:?} is unresolved", item_id),
-                ),
+            self.dcx().span_delayed_bug(
                 span,
-                delegation,
+                format!("LoweringContext: the delegation {:?} is unresolved", item_id),
             );
+
+            return self.generate_delegation_error(span, delegation);
         };
 
         match sig_id {
@@ -172,7 +170,7 @@ impl<'hir, R: ResolverAstLoweringExt<'hir>> LoweringContext<'_, 'hir, R> {
 
                 DelegationResults { body_id, sig, ident, generics }
             }
-            Err(err) => self.generate_delegation_error(err, span, delegation),
+            Err(_) => self.generate_delegation_error(span, delegation),
         }
     }
 
@@ -604,7 +602,6 @@ impl<'hir, R: ResolverAstLoweringExt<'hir>> LoweringContext<'_, 'hir, R> {
 
     fn generate_delegation_error(
         &mut self,
-        err: ErrorGuaranteed,
         span: Span,
         delegation: &Delegation,
     ) -> DelegationResults<'hir> {
@@ -622,36 +619,35 @@ impl<'hir, R: ResolverAstLoweringExt<'hir>> LoweringContext<'_, 'hir, R> {
         let ident = self.lower_ident(delegation.ident);
 
         let body_id = self.lower_body(|this| {
-            let body_expr = match delegation.body.as_ref() {
-                Some(box block) => {
-                    // Generates a block when we failed to resolve delegation, where a target expression is its only statement,
-                    // thus there will be no ICEs on further stages of analysis (see #144594)
+            let path = this.lower_qpath(
+                delegation.id,
+                &delegation.qself,
+                &delegation.path,
+                ParamMode::Optional,
+                AllowReturnTypeNotation::No,
+                ImplTraitContext::Disallowed(ImplTraitPosition::Path),
+                None,
+            );
 
-                    // As we generate a void function we want to convert target expression to statement to avoid additional
-                    // errors, such as mismatched return type
-                    let stmts = this.arena.alloc_from_iter([hir::Stmt {
-                        hir_id: this.next_id(),
-                        kind: rustc_hir::StmtKind::Semi(
-                            this.arena.alloc(this.lower_target_expr(block)),
-                        ),
-                        span,
-                    }]);
-
-                    let block = this.arena.alloc(hir::Block {
-                        stmts,
-                        expr: None,
-                        hir_id: this.next_id(),
-                        rules: hir::BlockCheckMode::DefaultBlock,
-                        span,
-                        targeted_by_break: false,
-                    });
-
-                    hir::ExprKind::Block(block, None)
-                }
-                None => hir::ExprKind::Err(err),
+            let callee_path = this.arena.alloc(this.mk_expr(hir::ExprKind::Path(path), span));
+            let args = if let Some(box block) = delegation.body.as_ref() {
+                this.arena.alloc_slice(&[this.lower_target_expr(block)])
+            } else {
+                &mut []
             };
 
-            (&[], this.mk_expr(body_expr, span))
+            let call = this.arena.alloc(this.mk_expr(hir::ExprKind::Call(callee_path, args), span));
+
+            let block = this.arena.alloc(hir::Block {
+                stmts: &[],
+                expr: Some(call),
+                hir_id: this.next_id(),
+                rules: hir::BlockCheckMode::DefaultBlock,
+                span,
+                targeted_by_break: false,
+            });
+
+            (&[], this.mk_expr(hir::ExprKind::Block(block, None), span))
         });
 
         let generics = hir::Generics::empty();

--- a/tests/ui/delegation/duplicate-definition-inside-trait-impl.rs
+++ b/tests/ui/delegation/duplicate-definition-inside-trait-impl.rs
@@ -18,6 +18,8 @@ impl Trait for S {
     reuse to_reuse::foo { self }
     reuse Trait::foo;
     //~^ ERROR  duplicate definitions with name `foo`
+    //~| ERROR: this function takes 1 argument but 0 arguments were supplied
+    //~| ERROR: mismatched types
 }
 
 fn main() {}

--- a/tests/ui/delegation/duplicate-definition-inside-trait-impl.stderr
+++ b/tests/ui/delegation/duplicate-definition-inside-trait-impl.stderr
@@ -9,6 +9,32 @@ LL |     reuse to_reuse::foo { self }
 LL |     reuse Trait::foo;
    |     ^^^^^^^^^^^^^^^^^ duplicate definition
 
-error: aborting due to 1 previous error
+error[E0061]: this function takes 1 argument but 0 arguments were supplied
+  --> $DIR/duplicate-definition-inside-trait-impl.rs:19:18
+   |
+LL |     reuse Trait::foo;
+   |                  ^^^ argument #1 of type `&_` is missing
+   |
+note: method defined here
+  --> $DIR/duplicate-definition-inside-trait-impl.rs:5:8
+   |
+LL |     fn foo(&self) -> u32 { 0 }
+   |        ^^^ -----
+help: provide the argument
+   |
+LL |     reuse Trait::foo(/* value */);
+   |                     +++++++++++++
 
-For more information about this error, try `rustc --explain E0201`.
+error[E0308]: mismatched types
+  --> $DIR/duplicate-definition-inside-trait-impl.rs:19:18
+   |
+LL |     reuse Trait::foo;
+   |                  ^^^- help: consider using a semicolon here: `;`
+   |                  |
+   |                  expected `()`, found `u32`
+   |                  expected `()` because of default return type
+
+error: aborting due to 3 previous errors
+
+Some errors have detailed explanations: E0061, E0201, E0308.
+For more information about an error, try `rustc --explain E0061`.

--- a/tests/ui/delegation/glob-glob-conflict.rs
+++ b/tests/ui/delegation/glob-glob-conflict.rs
@@ -3,9 +3,13 @@
 
 trait Trait1 {
     fn method(&self) -> u8;
+    //~^ ERROR: this function takes 1 argument but 0 arguments were supplied
+    //~| ERROR: mismatched types
 }
 trait Trait2 {
     fn method(&self) -> u8;
+    //~^ ERROR: this function takes 1 argument but 0 arguments were supplied
+    //~| ERROR: mismatched types
 }
 trait Trait {
     fn method(&self) -> u8;

--- a/tests/ui/delegation/glob-glob-conflict.stderr
+++ b/tests/ui/delegation/glob-glob-conflict.stderr
@@ -1,5 +1,5 @@
 error[E0201]: duplicate definitions with name `method`:
-  --> $DIR/glob-glob-conflict.rs:26:5
+  --> $DIR/glob-glob-conflict.rs:30:5
    |
 LL |     fn method(&self) -> u8;
    |     ----------------------- item in trait
@@ -10,7 +10,7 @@ LL |     reuse Trait2::*;
    |     ^^^^^^^^^^^^^^^^ duplicate definition
 
 error[E0201]: duplicate definitions with name `method`:
-  --> $DIR/glob-glob-conflict.rs:30:5
+  --> $DIR/glob-glob-conflict.rs:34:5
    |
 LL |     fn method(&self) -> u8;
    |     ----------------------- item in trait
@@ -20,6 +20,57 @@ LL |     reuse Trait1::*;
 LL |     reuse Trait1::*;
    |     ^^^^^^^^^^^^^^^^ duplicate definition
 
-error: aborting due to 2 previous errors
+error[E0061]: this function takes 1 argument but 0 arguments were supplied
+  --> $DIR/glob-glob-conflict.rs:10:8
+   |
+LL |     fn method(&self) -> u8;
+   |        ^^^^^^ argument #1 of type `&_` is missing
+   |
+note: method defined here
+  --> $DIR/glob-glob-conflict.rs:10:8
+   |
+LL |     fn method(&self) -> u8;
+   |        ^^^^^^  ----
+help: provide the argument
+   |
+LL |     fn method(/* value */)(&self) -> u8;
+   |              +++++++++++++
 
-For more information about this error, try `rustc --explain E0201`.
+error[E0308]: mismatched types
+  --> $DIR/glob-glob-conflict.rs:10:8
+   |
+LL |     fn method(&self) -> u8;
+   |        ^^^^^^- help: consider using a semicolon here: `;`
+   |        |
+   |        expected `()`, found `u8`
+   |        expected `()` because of default return type
+
+error[E0061]: this function takes 1 argument but 0 arguments were supplied
+  --> $DIR/glob-glob-conflict.rs:5:8
+   |
+LL |     fn method(&self) -> u8;
+   |        ^^^^^^ argument #1 of type `&_` is missing
+   |
+note: method defined here
+  --> $DIR/glob-glob-conflict.rs:5:8
+   |
+LL |     fn method(&self) -> u8;
+   |        ^^^^^^  ----
+help: provide the argument
+   |
+LL |     fn method(/* value */)(&self) -> u8;
+   |              +++++++++++++
+
+error[E0308]: mismatched types
+  --> $DIR/glob-glob-conflict.rs:5:8
+   |
+LL |     fn method(&self) -> u8;
+   |        ^^^^^^- help: consider using a semicolon here: `;`
+   |        |
+   |        expected `()`, found `u8`
+   |        expected `()` because of default return type
+
+error: aborting due to 6 previous errors
+
+Some errors have detailed explanations: E0061, E0201, E0308.
+For more information about an error, try `rustc --explain E0061`.

--- a/tests/ui/delegation/ice-issue-124347.rs
+++ b/tests/ui/delegation/ice-issue-124347.rs
@@ -4,9 +4,11 @@
 trait Trait {
     reuse Trait::foo { &self.0 }
     //~^ ERROR failed to resolve delegation callee
+    //~| ERROR: this function takes 0 arguments but 1 argument was supplied
 }
 
 reuse foo;
 //~^ ERROR failed to resolve delegation callee
+//~| WARN: function cannot return without recursing
 
 fn main() {}

--- a/tests/ui/delegation/ice-issue-124347.stderr
+++ b/tests/ui/delegation/ice-issue-124347.stderr
@@ -5,10 +5,40 @@ LL |     reuse Trait::foo { &self.0 }
    |                  ^^^
 
 error: failed to resolve delegation callee
-  --> $DIR/ice-issue-124347.rs:9:7
+  --> $DIR/ice-issue-124347.rs:10:7
    |
 LL | reuse foo;
    |       ^^^
 
-error: aborting due to 2 previous errors
+error[E0061]: this function takes 0 arguments but 1 argument was supplied
+  --> $DIR/ice-issue-124347.rs:5:18
+   |
+LL |     reuse Trait::foo { &self.0 }
+   |                  ^^^   ------- unexpected argument
+   |
+note: associated function defined here
+  --> $DIR/ice-issue-124347.rs:5:18
+   |
+LL |     reuse Trait::foo { &self.0 }
+   |                  ^^^
+help: remove the extra argument
+   |
+LL -     reuse Trait::foo { &self.0 }
+LL +     reuse Trait::fo&self.0 }
+   |
 
+warning: function cannot return without recursing
+  --> $DIR/ice-issue-124347.rs:10:7
+   |
+LL | reuse foo;
+   |       ^^^
+   |       |
+   |       cannot return without recursing
+   |       recursive call site
+   |
+   = help: a `loop` may express intention better if this is on purpose
+   = note: `#[warn(unconditional_recursion)]` on by default
+
+error: aborting due to 3 previous errors; 1 warning emitted
+
+For more information about this error, try `rustc --explain E0061`.

--- a/tests/ui/delegation/recursive-delegation-errors.rs
+++ b/tests/ui/delegation/recursive-delegation-errors.rs
@@ -5,6 +5,7 @@
 mod first_mod {
     reuse foo;
     //~^ ERROR failed to resolve delegation callee
+    //~| WARN: function cannot return without recursing
 }
 
 mod second_mod {
@@ -33,8 +34,10 @@ mod fourth_mod {
     trait Trait {
         reuse Trait::foo as bar;
         //~^ ERROR encountered a cycle during delegation signature resolution
+        //~| ERROR: type annotations needed
         reuse Trait::bar as foo;
         //~^ ERROR encountered a cycle during delegation signature resolution
+        //~| ERROR: type annotations needed
     }
 }
 
@@ -48,6 +51,9 @@ mod fifth_mod {
         //~^ ERROR encountered a cycle during delegation signature resolution
         //~| ERROR encountered a cycle during delegation signature resolution
         //~| ERROR encountered a cycle during delegation signature resolution
+        //~| ERROR: type annotations needed
+        //~| ERROR: type annotations needed
+        //~| ERROR: type annotations needed
     }
 }
 

--- a/tests/ui/delegation/recursive-delegation-errors.stderr
+++ b/tests/ui/delegation/recursive-delegation-errors.stderr
@@ -5,94 +5,147 @@ LL |     reuse foo;
    |           ^^^
 
 error: encountered a cycle during delegation signature resolution
-  --> $DIR/recursive-delegation-errors.rs:11:11
+  --> $DIR/recursive-delegation-errors.rs:12:11
    |
 LL |     reuse foo as bar;
    |           ^^^
 
 error: encountered a cycle during delegation signature resolution
-  --> $DIR/recursive-delegation-errors.rs:13:11
+  --> $DIR/recursive-delegation-errors.rs:14:11
    |
 LL |     reuse bar as foo;
    |           ^^^
 
 error: encountered a cycle during delegation signature resolution
-  --> $DIR/recursive-delegation-errors.rs:18:11
+  --> $DIR/recursive-delegation-errors.rs:19:11
    |
 LL |     reuse foo as foo1;
    |           ^^^
 
 error: encountered a cycle during delegation signature resolution
-  --> $DIR/recursive-delegation-errors.rs:20:11
+  --> $DIR/recursive-delegation-errors.rs:21:11
    |
 LL |     reuse foo1 as foo2;
    |           ^^^^
 
 error: encountered a cycle during delegation signature resolution
-  --> $DIR/recursive-delegation-errors.rs:22:11
+  --> $DIR/recursive-delegation-errors.rs:23:11
    |
 LL |     reuse foo2 as foo3;
    |           ^^^^
 
 error: encountered a cycle during delegation signature resolution
-  --> $DIR/recursive-delegation-errors.rs:24:11
+  --> $DIR/recursive-delegation-errors.rs:25:11
    |
 LL |     reuse foo3 as foo4;
    |           ^^^^
 
 error: encountered a cycle during delegation signature resolution
-  --> $DIR/recursive-delegation-errors.rs:26:11
+  --> $DIR/recursive-delegation-errors.rs:27:11
    |
 LL |     reuse foo4 as foo5;
    |           ^^^^
 
 error: encountered a cycle during delegation signature resolution
-  --> $DIR/recursive-delegation-errors.rs:28:11
+  --> $DIR/recursive-delegation-errors.rs:29:11
    |
 LL |     reuse foo5 as foo;
    |           ^^^^
 
 error: encountered a cycle during delegation signature resolution
-  --> $DIR/recursive-delegation-errors.rs:34:22
+  --> $DIR/recursive-delegation-errors.rs:35:22
    |
 LL |         reuse Trait::foo as bar;
    |                      ^^^
 
 error: encountered a cycle during delegation signature resolution
-  --> $DIR/recursive-delegation-errors.rs:36:22
+  --> $DIR/recursive-delegation-errors.rs:38:22
    |
 LL |         reuse Trait::bar as foo;
    |                      ^^^
 
 error: encountered a cycle during delegation signature resolution
-  --> $DIR/recursive-delegation-errors.rs:42:30
+  --> $DIR/recursive-delegation-errors.rs:45:30
    |
 LL |     reuse super::fifth_mod::{bar as foo, foo as bar};
    |                              ^^^
 
 error: encountered a cycle during delegation signature resolution
-  --> $DIR/recursive-delegation-errors.rs:42:42
+  --> $DIR/recursive-delegation-errors.rs:45:42
    |
 LL |     reuse super::fifth_mod::{bar as foo, foo as bar};
    |                                          ^^^
 
 error: encountered a cycle during delegation signature resolution
-  --> $DIR/recursive-delegation-errors.rs:47:27
+  --> $DIR/recursive-delegation-errors.rs:50:27
    |
 LL |         reuse GlobReuse::{foo as bar, bar as goo, goo as foo};
    |                           ^^^
 
 error: encountered a cycle during delegation signature resolution
-  --> $DIR/recursive-delegation-errors.rs:47:39
+  --> $DIR/recursive-delegation-errors.rs:50:39
    |
 LL |         reuse GlobReuse::{foo as bar, bar as goo, goo as foo};
    |                                       ^^^
 
 error: encountered a cycle during delegation signature resolution
-  --> $DIR/recursive-delegation-errors.rs:47:51
+  --> $DIR/recursive-delegation-errors.rs:50:51
    |
 LL |         reuse GlobReuse::{foo as bar, bar as goo, goo as foo};
    |                                                   ^^^
 
-error: aborting due to 16 previous errors
+error[E0283]: type annotations needed
+  --> $DIR/recursive-delegation-errors.rs:35:22
+   |
+LL |         reuse Trait::foo as bar;
+   |                      ^^^ cannot infer type
+   |
+   = note: the type must implement `fourth_mod::Trait`
 
+error[E0283]: type annotations needed
+  --> $DIR/recursive-delegation-errors.rs:38:22
+   |
+LL |         reuse Trait::bar as foo;
+   |                      ^^^ cannot infer type
+   |
+   = note: the type must implement `fourth_mod::Trait`
+
+error[E0283]: type annotations needed
+  --> $DIR/recursive-delegation-errors.rs:50:27
+   |
+LL |         reuse GlobReuse::{foo as bar, bar as goo, goo as foo};
+   |                           ^^^ cannot infer type
+   |
+   = note: the type must implement `GlobReuse`
+
+error[E0283]: type annotations needed
+  --> $DIR/recursive-delegation-errors.rs:50:39
+   |
+LL |         reuse GlobReuse::{foo as bar, bar as goo, goo as foo};
+   |                                       ^^^ cannot infer type
+   |
+   = note: the type must implement `GlobReuse`
+
+error[E0283]: type annotations needed
+  --> $DIR/recursive-delegation-errors.rs:50:51
+   |
+LL |         reuse GlobReuse::{foo as bar, bar as goo, goo as foo};
+   |                                                   ^^^ cannot infer type
+   |
+   = note: the type must implement `GlobReuse`
+
+warning: function cannot return without recursing
+  --> $DIR/recursive-delegation-errors.rs:6:11
+   |
+LL |     reuse foo;
+   |           ^^^
+   |           |
+   |           cannot return without recursing
+   |           recursive call site
+   |
+   = help: a `loop` may express intention better if this is on purpose
+   = note: `#[warn(unconditional_recursion)]` on by default
+
+error: aborting due to 21 previous errors; 1 warning emitted
+
+For more information about this error, try `rustc --explain E0283`.

--- a/tests/ui/delegation/unlowered-path-ice-154820.rs
+++ b/tests/ui/delegation/unlowered-path-ice-154820.rs
@@ -1,0 +1,12 @@
+#![feature(fn_delegation)]
+#![allow(incomplete_features)]
+
+reuse foo:: < { //~ ERROR: failed to resolve delegation callee
+  //~^ ERROR: function takes 0 generic arguments but 1 generic argument was supplied
+    fn foo() {}
+    reuse foo;
+    //~^ ERROR: the name `foo` is defined multiple times
+  }
+  >;
+
+fn main() {}

--- a/tests/ui/delegation/unlowered-path-ice-154820.stderr
+++ b/tests/ui/delegation/unlowered-path-ice-154820.stderr
@@ -1,0 +1,40 @@
+error[E0428]: the name `foo` is defined multiple times
+  --> $DIR/unlowered-path-ice-154820.rs:7:5
+   |
+LL |     fn foo() {}
+   |     -------- previous definition of the value `foo` here
+LL |     reuse foo;
+   |     ^^^^^^^^^^ `foo` redefined here
+   |
+   = note: `foo` must be defined only once in the value namespace of this block
+
+error: failed to resolve delegation callee
+  --> $DIR/unlowered-path-ice-154820.rs:4:7
+   |
+LL | reuse foo:: < {
+   |       ^^^
+
+error[E0107]: function takes 0 generic arguments but 1 generic argument was supplied
+  --> $DIR/unlowered-path-ice-154820.rs:4:7
+   |
+LL |   reuse foo:: < {
+   |  _______^^^-
+   | |       |
+   | |       expected 0 generic arguments
+LL | |
+LL | |     fn foo() {}
+LL | |     reuse foo;
+...  |
+LL | |   >;
+   | |___- help: remove the unnecessary generics
+   |
+note: function defined here, with 0 generic parameters
+  --> $DIR/unlowered-path-ice-154820.rs:4:7
+   |
+LL | reuse foo:: < {
+   |       ^^^
+
+error: aborting due to 3 previous errors
+
+Some errors have detailed explanations: E0107, E0428.
+For more information about an error, try `rustc --explain E0107`.

